lib/mongoid/scroll/base64_encoded_cursor.rb

```diff
@@ -0,0 +1,43 @@
+require 'base64'
+require 'json'
+
+module Mongoid
+  module Scroll
+    # Allows to serializer/deserialize the cursor using RFC 4648
+    class Base64EncodedCursor < Cursor
+      class << self
+        def from_cursor(cursor)
+          base64_encoded_cursor = new(nil,
+                                      field_type: cursor.field_type,
+                                      field_name: cursor.field_name,
+                                      direction: cursor.direction,
+                                      include_current: cursor.include_current)
+          base64_encoded_cursor.value = cursor.value
+          base64_encoded_cursor.tiebreak_id = cursor.tiebreak_id
+          base64_encoded_cursor
+        end
+      end
+
+      def to_s
+        Base64.strict_encode64({ value: super, field_type: field_type, field_name: field_name, direction: direction, include_current: include_current }.to_json)
+      end
+
+      private
+
+      def parse(value)
+        return unless value
+
+        begin
+          config_hash = ::JSON.parse(::Base64.strict_decode64(value))
+        rescue
+          raise Mongoid::Scroll::Errors::InvalidCursorError.new(cursor: value)
+        end
+        @field_type = config_hash['field_type']
+        @field_name = config_hash['field_name']
+        @direction = config_hash['direction']
+        @include_current = config_hash['include_current']
+        super(config_hash['value'])
```

Comment:
Pass field options down, this way you can remove the check for class in `extract_field_options`.

```
super config_hash['value'], config_hash
```

Reply:
`super` is calling `Cursor#parse` which only accepts a `value` argument, not an extra `options` one. It's not this method which is raising (because of `Cursor.extract_field_options`) if the `:field_name`/`:field_name` options are missing, it's the `initialize` method.

If we always raise in `Cursor.extract_field_options` (removing the `elsif self == Mongoid::Scroll::Cursor` that I added), we won't be able to create a `Base64EncodedCursor` without option (`Mongoid::Scroll::Base64EncodedCursor.new(base64_string)` will raise).

Reply:
Ok I see. I think we're trying to reuse too much of the base cursor which attempts to join things with `:` and what not, all while it is completely unnecessary for a base64 cursor which is happy to just base64(hash). How about we don't call super in parse at all as you suggested before? Override the constructor and to_s?

Maybe we should have a `Base64Cursor`, a `Cursor` (string cursor) and a `Base` for common parts?